src/main/java/com/treasuredata/bigdam/log/MetricMonitor.java

```diff
@@ -0,0 +1,170 @@
+
+    private static final long SLEEP_INTERVAL_MSEC = 800L;
+    private static final long MONITOR_THREAD_STOP_TIMEOUT = 2000L; // sleep interva * 2 + alpha
+
+    public MetricMonitor(
+            final Log logger,
+            final String tagPrefixMetric,
+            final String tagPrefixRawMetric,
+            final String metricFieldName,
+            final int metricIntervalSeconds
+    )
+    {
+        this.logger = logger;
+
+        this.tagPrefixMetric = tagPrefixMetric;
+        this.tagPrefixRawMetric = tagPrefixRawMetric;
+        this.metricFieldName = metricFieldName;
+
+        this.metricIntervalSeconds = metricIntervalSeconds;
+        this.sleepIntervalMilliSeconds = SLEEP_INTERVAL_MSEC;
+
+        this.metricProducers = new ArrayList<>();
+        this.rawMetricProducers = new ArrayList<>();
+        this.metricProducersComplex = new ArrayList<>();
+        this.rawMetricProducersComplex = new ArrayList<>();
+
+        this.running = new AtomicBoolean(false);
+    }
+
+    public void addMetricProducer(final MetricProducer producer)
+    {
+        metricProducers.add(producer);
+    }
+
+    public void addMetricProducer(final ComplexMetricProducer producer)
+    {
+        metricProducersComplex.add(producer);
+    }
+
+    public void addRawMetricProducer(final MetricProducer producer)
+    {
+        rawMetricProducers.add(producer);
+    }
+
+    public void addRawMetricProducer(final ComplexMetricProducer producer)
+    {
+        rawMetricProducersComplex.add(producer);
+    }
+
+    // only for MetricMonitorTest
+    void setSleepInterval(final long sleepInterval)
+    {
+        this.sleepIntervalMilliSeconds = sleepInterval;
+    }
+
+    public void start()
+    {
+        running.set(true);
+        monitorThread = new Thread(this::loop);
+        monitorThread.start();
+    }
+
+    public void stop()
+    {
+        running.set(false);
+        try {
+            monitorThread.join(MONITOR_THREAD_STOP_TIMEOUT);
+            if (monitorThread.isAlive()) {
+                monitorThread.interrupt();
+                monitorThread.join(MONITOR_THREAD_STOP_TIMEOUT);
+            }
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt(); // preserve interrupted status
+        }
+    }
+
+    private void loop()
+    {
+        long metricInterval = metricIntervalSeconds * 1_000_000_000L;
+
+        long nextTick = Clock.now() + metricInterval;
+        while (running.get()) {
+            try {
+                Thread.sleep(sleepIntervalMilliSeconds);
+                if (running.get() && Clock.now() >= nextTick) {
+                    nextTick += metricInterval; // to set nextTick even when run() throws exceptions
+                    run();
+                }
+            }
+            catch (Throwable e) {
+                logger.error(String.format("MetricMonitor got an error %s: %s", e.getClass().getName(), e.getMessage()), e);
```

Comment:
Log.error doesn't support format string like `logger.error("MetricMonitor got an error {}: {}")`?

Reply:
It's not supported now, because that argument message is also passed to Fluency, and it's a bit hard to implement such formatter in Fluency too.